README.md
[![Get Support](https://xscode.com/assets/promo-banner.svg)](https://xscode.com/eshaan7/Flask-Shell2HTTP)

_For urgent issues and priority support, visit [https://xscode.com/eshaan7/flask-shell2http](https://xscode.com/eshaan7/flask-shell2http)._

# Flask-Shell2HTTP

[![flask-shell2http on pypi](https://img.shields.io/pypi/v/flask-shell2http)](https://pypi.org/project/Flask-Shell2HTTP/)
[![Build Status](https://github.com/Eshaan7/flask-shell2http/workflows/Linter%20&%20Tests/badge.svg?branch=master)](https://github.com/Eshaan7/flask-shell2http/actions?query=workflow%3A%22Linter+%26+Tests%22)
[![codecov](https://codecov.io/gh/Eshaan7/Flask-Shell2HTTP/branch/master/graph/badge.svg?token=UQ43PYQPMR)](https://codecov.io/gh/Eshaan7/flask-shell2http/)
[![CodeFactor](https://www.codefactor.io/repository/github/eshaan7/flask-shell2http/badge)](https://www.codefactor.io/repository/github/eshaan7/flask-shell2http)
<a href="https://lgtm.com/projects/g/Eshaan7/Flask-Shell2HTTP/context:python">
<img alt="Language grade: Python" src="https://img.shields.io/lgtm/grade/python/g/Eshaan7/Flask-Shell2HTTP.svg?logo=lgtm&logoWidth=18"/>
</a>

A minimalist [Flask](https://github.com/pallets/flask) extension that serves as a RESTful/HTTP wrapper for python's subprocess API.

- **Convert any command-line tool into a REST API service.**
- Execute pre-defined shell commands asynchronously and securely via flask's endpoints with dynamic arguments, file upload, callback function capabilities.
- Designed for binary to binary/HTTP communication, development, prototyping, remote control and [more](https://flask-shell2http.readthedocs.io/en/stable/Examples.html).

## Use Cases

- Set a script that runs on a succesful POST request to an endpoint of your choice. See [Example code](examples/run_script.py).
- Map a base command to an endpoint and pass dynamic arguments to it. See [Example code](examples/basic.py).
- Can also process multiple uploaded files in one command. See [Example code](examples/multiple_files.py).
- This is useful for internal docker-to-docker communications if you have different binaries distributed in micro-containers. See [real-life example](https://github.com/intelowlproject/IntelOwl/blob/master/integrations/static_analyzers/app.py).
- You can define a callback function/ use signals to listen for process completion. See [Example code](examples/with_callback.py).
  - Maybe want to pass some additional context to the callback function ?
  - Maybe intercept on completion and update the result ? See [Example code](examples/custom_save_fn.py)
- You can also apply [View Decorators](https://flask.palletsprojects.com/en/1.1.x/patterns/viewdecorators/) to the exposed endpoint. See [Example code](examples/with_decorators.py).

> Note: This extension is primarily meant for executing long-running
> shell commands/scripts (like nmap, code-analysis' tools) in background from an HTTP request and getting the result at a later time.

## Documentation

[![Documentation Status](https://readthedocs.org/projects/flask-shell2http/badge/?version=latest)](https://flask-shell2http.readthedocs.io/en/latest/?badge=latest)

- Read the [Quickstart](https://flask-shell2http.readthedocs.io/en/stable/Quickstart.html) from the [documentation](https://flask-shell2http.readthedocs.io/) to get started!
- I also highly recommend the [Examples](https://flask-shell2http.readthedocs.io/en/stable/Examples.html) section.
- [CHANGELOG](https://github.com/eshaan7/Flask-Shell2HTTP/blob/master/.github/CHANGELOG.md).

## Quick Start

##### Dependencies

- Python: `>=v3.6`
- [Flask](https://pypi.org/project/Flask/)
- [Flask-Executor](https://pypi.org/project/Flask-Executor)

##### Installation

```bash
$ pip install flask flask_shell2http
```

##### Example Program

Create a file called `app.py`.

```python
from flask import Flask
from flask_executor import Executor
from flask_shell2http import Shell2HTTP

# Flask application instance
app = Flask(__name__)

executor = Executor(app)
shell2http = Shell2HTTP(app=app, executor=executor, base_url_prefix="/commands/")

def my_callback_fn(context, future):
  # optional user-defined callback function
  print(context, future.result())

shell2http.register_command(endpoint="saythis", command_name="echo", callback_fn=my_callback_fn, decorators=[])
```

Run the application server with, `$ flask run -p 4000`.

With <10 lines of code, we succesfully mapped the shell command `echo` to the endpoint `/commands/saythis`.

##### Making HTTP calls

This section demonstrates how we can now call/ execute commands over HTTP that we just mapped in the [example](#example-program) above.

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"args": ["Hello", "World!"]}' http://localhost:4000/commands/saythis
```

<details><summary>or using python's requests module,</summary>

```python
# You can also add a timeout if you want, default value is 3600 seconds
data = {"args": ["Hello", "World!"], "timeout": 60}
resp = requests.post("http://localhost:4000/commands/saythis", json=data)
print("Result:", resp.json())
```

</details>

> Note: You can see the JSON schema for the POST request [here](https://github.com/Eshaan7/Flask-Shell2HTTP/blob/master/post-request-schema.json).

returns JSON,

```json
{
  "key": "ddbe0a94",
  "result_url": "http://localhost:4000/commands/saythis?key=ddbe0a94&wait=false",
  "status": "running"
}
```

Then using this `key` you can query for the result or just by going to the `result_url`,

```bash
$ curl http://localhost:4000/commands/saythis?key=ddbe0a94&wait=true # wait=true so we do not have to poll
```

Returns result in JSON,

```json
{
  "report": "Hello World!\n",
  "key": "ddbe0a94",
  "start_time": 1593019807.7754705,
  "end_time": 1593019807.782958,
  "process_time": 0.00748753547668457,
  "returncode": 0,
  "error": null
}
```

## Inspiration

This was initially made to integrate various command-line tools easily with [Intel Owl](https://github.com/intelowlproject/IntelOwl), which I am working on as part of Google Summer of Code.

The name was inspired by the awesome folks over at [msoap/shell2http](https://github.com/msoap/shell2http).


.github/CHANGELOG.md
# Changelog

**[Get it on PyPi](https://pypi.org/project/Flask-Shell2HTTP/)**

## [v1.9.1](https://github.com/eshaan7/Flask-Shell2HTTP/releases/tag/v1.9.1)

Added support for Flask version >=2.1.0 .

- Fix import errors. [[Issue 42](https://github.com/eshaan7/Flask-Shell2HTTP/issues/42)].
- Fix dependency conflicts in `requirements.dev.txt` file.

## [v1.9.0](https://github.com/eshaan7/Flask-Shell2HTTP/releases/tag/v1.9.0)

- It's now possible to request deletion/cancellation of jobs. [[Issue 38](https://github.com/eshaan7/Flask-Shell2HTTP/issues/38)].

  Say for example, you register a command with the endpoint `/myendpoint` then the different HTTP operations supported are:

  - `POST /myendpoint` to create a job and get unique key (same as earlier)
  - `GET /myendpoint?key={key}` to request result of a job (same as earlier)
  - `DELETE /myendpoint?key={key}` to request cancellation/deletion of a job. (new)

## [v1.8.0](https://github.com/eshaan7/Flask-Shell2HTTP/releases/tag/v1.8.0)

- Allow `&wait=[false|true]` query parameter in `GET` request. Use `wait=true` when you don't wish to HTTP poll and want the result in a single request only.

## [v1.7.0](https://github.com/eshaan7/Flask-Shell2HTTP/releases/tag/v1.7.0)

**For you:**

- Deps: Support for both Flask version 1.x and 2.x.
- Feature: The `key` and `result_url` attributes are returned in response even if error is raised (if and when applicable) (See [#25](https://github.com/eshaan7/Flask-Shell2HTTP/issues/25)).
- Docs: Add info about `force_unique_key` option to quickstart guide.

**Internal:**

- Much better and improved test cases via tox matrix for both major flask versions, 1.x and 2.x.
- Much better overall type hinting.

## [v1.6.0](https://github.com/eshaan7/Flask-Shell2HTTP/releases/tag/v1.6.0)

Added support for a new parameter `force_unique_key` in the POST request.

```json
    "force_unique_key": {
      "type": "boolean",
      "title": "Flag to enable/disable internal rate limiting mechanism",
      "description": "By default, the key is the SHA1 sum of the command + args POSTed to the API. This is done as a rate limiting measure so as to prevent multiple jobs with same parameters, if one such job is already running. If force_unique_key is set to true, the API will bypass this default behaviour and a psuedorandom key will be returned instead",
      "default": false
    }
```

See [post-request-options configuration](https://flask-shell2http.readthedocs.io/en/latest/Configuration.html#post-request-options) in docs for more info.

_For prior versions, see directly [here](https://github.com/eshaan7/Flask-Shell2HTTP/releases)._


examples/basic.py
# system imports
import logging
import sys

# web imports
from flask import Flask
from flask_executor import Executor
from flask_shell2http import Shell2HTTP

# Flask application instance
app = Flask(__name__)

# Logging
formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
logger = logging.getLogger("flask_shell2http")
handler = logging.StreamHandler(sys.stdout)
handler.setFormatter(formatter)
logger.addHandler(handler)
logger.setLevel(logging.INFO)


# application factory
executor = Executor(app)
shell2http = Shell2HTTP(app, executor, base_url_prefix="/cmd/")

ENDPOINT = "echo"
shell2http.register_command(endpoint=ENDPOINT, command_name=ENDPOINT)


# Test Runner
if __name__ == "__main__":
    app.testing = True
    c = app.test_client()
    """
    The final executed command becomes:
    ```bash
    $ echo hello world
    ```
    """
    # make new request for a command with arguments
    uri = f"/cmd/{ENDPOINT}"
    # timeout in seconds, default value is 3600
    # force_unique_key disables rate-limiting
    data = {"args": ["hello", "world"], "timeout": 60, "force_unique_key": True}
    resp1 = c.post(uri, json=data).get_json()
    print(resp1)
    # fetch result
    result_url = resp1["result_url"].replace("wait=false", "wait=true")
    resp2 = c.get(result_url).get_json()
    print(resp2)
